src/MCPServer/lib/package.py

```diff
@@ -0,0 +1,391 @@
+    name = unicodeToStr(name)
+    type_ = unicodeToStr(type_ if type_ is not None else 'standard')
+    if type_ not in PACKAGE_TYPE_STARTING_POINTS:
+        raise ValueError('Unexpected type of package provided.')
+    # Create Transfer object.
+    kwargs = {'uuid': str(uuid4())}
+    if accession is not None:
+        kwargs['accessionid'] = unicodeToStr(accession)
+    if metadata_set_id is not None:
+        try:
+            kwargs['transfermetadatasetrow'] = \
+                TransferMetadataSet.objects.get(id=metadata_set_id)
+        except TransferMetadataSet.DoesNotExist:
+            pass
+    transfer = Transfer.objects.create(**kwargs)
+    # We want to return as soon as possible. Do the rest of the work in the
+    # background so we don't make the caller wait. We persist the job handle so
+    # we can use it later to attempt to know the status of the background job.
+    _submit_create_package_internal_bg_job(transfer.pk, name, type_, path)
+    return transfer
+
+
+def _submit_create_package_internal_bg_job(id_, name, type_, path):
+    """Submit package creation background job to queue.
+
+    It returns the job handle assigned by Gearman.
+    """
+    payload = cPickle.dumps({
+        'id': id_, 'name': name, 'type': type_, 'path': path
+    })
+    gm_client = gearman.GearmanClient([django_settings.GEARMAN_SERVER])
+    request = gm_client.submit_job('MCPServerInternalPackageStartTransfer',
+                                   payload, background=True,
+                                   wait_until_complete=False)
+    gm_client.shutdown()
+    return request.job.handle
+
+
+def _create_temporary_directory():
+    """Create temporary directory in Archivematica Shared Directory.
+
+    This is useful if you want to do some work on the shared directory mainly
+    because you expect other actors to participate, e.g. Storage Service.
+    The caller is responsible for removing the directory when needed.
+    """
+    tmpdir = mkdtemp(dir=os.path.join(django_settings.SHARED_DIRECTORY, 'tmp'))
+    os.chmod(tmpdir, 0o770)  # Needs to be writeable by the SS.
+    return tmpdir
+
+
+def _capture_transfer_failure(fn):
+    """Detect errors during transfer/ingest."""
+    def wrap(*args, **kwargs):
+        err = None
+        try:
+            # Our decorated isn't expected to return anything.
+            fn(*args, **kwargs)
+        except Exception as e:
+            # The main purpose of this decorator is to update the Transfer with
+            # the new state (fail). If the Transfer does not exist we give up.
+            if isinstance(err, Transfer.DoesNotExist):
+                raise
+            err = e
+        else:
+            # No exceptions!
+            return
+        # At this point we know that the transfer has failed and we want to do
+        # our best effort to update the state without further interruptions.
+        try:
+            Transfer.objects.get(pk=args[0])
+
+            #
```

Comment:
I had some code here that was updating the `Transfer.state` FSM. We decided to think about it a bit further so I deleted it. I'm going to add a comment to make that clear. I wanted to leave the decorator in place because I thought it would be clearer to have it even if it's a no-op right now.